clang/lib/Driver/OffloadBundler.cpp

```diff
@@ -942,20 +942,28 @@ CompressedOffloadBundle::compress(const llvm::MemoryBuffer &Input,
       Input.getBuffer().size());
 
   llvm::compression::Format CompressionFormat;
+  int Level;
 
-  if (llvm::compression::zstd::isAvailable())
+  if (llvm::compression::zstd::isAvailable()) {
     CompressionFormat = llvm::compression::Format::Zstd;
-  else if (llvm::compression::zlib::isAvailable())
+    // Use a high zstd compress level by default for better size reduction.
+    const int DefaultZstdLevel = 20;
```

Comment:
What's the default compression level for zstd? 

It would be great if we could override the compression level. I'm somewhat reluctant to impose max compression level on everyone by default, without any way out, if it turns out to be a problem.

@MaskRay WDYT? 

Max compression level may be fine. If we produce enough stuff for compression to take long, compilation time itself will likely dwarf the compression time. For the small TUs, even slow compression may be fine.

@yxsamliu how long the compilation w/o compression takes in your benchmarks?

Reply:
> What's the default compression level for zstd?
> 
default compression level for zstd is 6

> It would be great if we could override the compression level. I'm somewhat reluctant to impose max compression level on everyone by default, without any way out, if it turns out to be a problem.
> 

I can add a clang-offload-bundler option -compress-level=n and a clang option --offload-compress-level=n to control the compression level.

> @MaskRay WDYT?
> 
> Max compression level may be fine. If we produce enough stuff for compression to take long, compilation time itself will likely dwarf the compression time. For the small TUs, even slow compression may be fine.
> 
> @yxsamliu how long the compilation w/o compression takes in your benchmarks?

For Blender 4.1, compiling kernels to bitcode for 6 GPU takes 30s. compression with zstd level 20 takes 2s.

Reply:
High levels decrease compression speed substantially. Have you checked that 20 hits a good sweet spot between 9 to 22 (max)?

Is OffloadBundler used more like a linker than a compile action where parallelism helps? If yes, lld/ELF/OutputSections.cpp has some code, which can be copied here. (Please don't share the code, I think some duplication in lld/ELF makes lld code easier to change)

Reply:
level 20 is a sweet spot for both compression rate and compression time, as shown in the following table for Blender 4.1 bundled bitcode for 6 GPU archs:

| Zstd Level | Size Before (bytes) | Size After (bytes) | Compression Rate | Compression Time (s) | Decompression Time (s) |
|------------|---------------------|--------------------|------------------|----------------------|------------------------|
| 6          | 68,459,756          | 32,612,291         | 2.10             | 0.8891               | 0.1809                 |
| 9          | 68,459,756          | 31,445,373         | 2.18             | 1.4200               | 0.1742                 |
| 15         | 68,459,756          | 28,063,493         | 2.44             | 9.7994               | 0.1712                 |
| 18         | 68,459,756          | 24,952,891         | 2.74             | 11.4201              | 0.1796                 |
| 19         | 68,459,756          | 24,690,733         | 2.77             | 13.4060              | 0.1820                 |
| 20         | 68,459,756          | 4,394,993          | 15.58            | 2.0946               | 0.1320                 |
| 21         | 68,459,756          | 4,394,399          | 15.59            | 2.1500               | 0.1318                 |
| 22         | 68,459,756          | 4,394,429          | 15.59            | 2.6635               | 0.1309                 |

Level 20 and level 19 has some compression parameter differences (https://github.com/facebook/zstd/blob/a58b48ef0e543980888a4d9d16c9072ff22135ca/lib/compress/clevels.h#L48 ) the meaning of these parameters are defined at https://github.com/facebook/zstd/blob/a58b48ef0e543980888a4d9d16c9072ff22135ca/lib/zstd.h#L1299. It seems either the largest match distance or fully searched segment makes the difference.

clang-offload-bundler just concatenates the binaries for different GPU arch's together. Parallelization does not help much, unless zstd can be parallelized.

Reply:
> compiling kernels to bitcode for 6 GPU takes 30s. compression with zstd level 20 takes 2s.

This looks acceptable for me.

> unless zstd can be parallelized.

zstd does support multithreaded compression, but enabling it would run into the same issue we had with enabling multi-threaded compilation -- it will interfere with the build system's idea of resource usage.